bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/internal/cache/TTSResult.java

```diff
@@ -0,0 +1,521 @@
+
+    /**
+     * Arbitrary chunk size. Small is less latency but more small calls and CPU load.
+     */
+    private static final int CHUNK_SIZE = 10000;
+
+    /**
+     * Take count of the number of {@link AudioStreamCacheWrapper} currently using this {@link TTSResult}
+     */
+    private Integer countAudioStreamClient = 0;
+
+    /**
+     * A unique key to identify the result
+     * (used to build the filename)
+     */
+    private final String key;
+
+    private String text;
+
+    // The inner AudioStream, with a supplier
+    @Nullable
+    private AudioStreamSupplier ttsAudioStreamSupplier;
+    @Nullable
+    private AudioStream ttsAudioStream;
+    @Nullable
+    private AudioFormat audioFormat;
+
+    // The cache files where the sound and its info are stored :
+    private File soundFile;
+    private File infoFile;
+    private long currentSize = 0;
+    private boolean completed;
+
+    @NonNullByDefault({}) // file channel could not be null when we read or write from the wrapper
+    private FileChannel fileChannel;
+    private Lock fileOperationLock = new ReentrantLock();
+
+    /**
+     * These references link {@link TTSResult} between each other, allowing LRU cache to work.
+     */
+    @Nullable
+    private TTSResult next = null;
+    @Nullable
+    private TTSResult prev = null;
+
+    /**
+     * This constructor is used when the file is fully cached on disk.
+     * The file on disk will provide the data, and the .info file will
+     * provide metadata about the format.
+     *
+     * @param cacheDirectory Where are the cached file stored
+     * @param key a unique key
+     */
+    public TTSResult(File cacheDirectory, String key) throws IOException {
+        this.key = key;
+        this.soundFile = new File(cacheDirectory, key + TTSLRUCacheImpl.SOUND_EXT);
+        this.infoFile = new File(cacheDirectory, key + TTSLRUCacheImpl.INFO_EXT);
+        this.ttsAudioStreamSupplier = null;
+        this.completed = true;
+        if (this.soundFile.length() == 0) {
+            throw new IOException("Sound cache file is empty. Throwing it");
+        }
+        this.currentSize = this.soundFile.length();
+
+        Properties properties = new Properties();
+        try (FileReader infoFileReader = new FileReader(this.infoFile)) {
+            properties.load(infoFileReader);
+            String containerS = properties.getProperty(PROPERTY_FORMAT_CONTAINER);
+            String bigEndianS = properties.getProperty(PROPERTY_FORMAT_BIGENDIAN);
+            String bitDepthS = properties.getProperty(PROPERTY_FORMAT_BITDEPTH);
+            String bitRateS = properties.getProperty(PROPERTY_FORMAT_BITRATE);
+            String channelsS = properties.getProperty(PROPERTY_FORMAT_CHANNELS);
+            String codecS = properties.getProperty(PROPERTY_FORMAT_CODEC);
+            String frequencyS = properties.getProperty(PROPERTY_FORMAT_FREQUENCY);
+            String textS = properties.getProperty(PROPERTY_TEXT);
```

Comment:
Wouldn't it make sense to create a `record` or `class` holding this information and leave the serialization/deserialization to `Gson`? Probably performance is not so much an issue here and you could simplify code to about two or three lines here.

Reply:
Interesting !
Sadly, deserializing record needs Gson 2.10, and we have Gson 2.9.1
Do you think it could be pushed for openHAB 2.4 ?

I will use a standard class for the moment.

Reply:
Probably when we switch to Karaf 4.4. class is fine for now.